### ISSUE TITLE
Unhandled Promise errors

### ISSUE THREAD

**Issue by timoxley** [paulmillr/es6-shim#311]
Currently, if there's no `.catch` errors will vanish mysteriously. Native promises in chrome  41.0.2249.0 (current canary) will print a message about unhandled promise errors:

![image](https://cloud.githubusercontent.com/assets/43438/5430004/891d6aa6-843d-11e4-8df2-2a924e714637.png)

but if the es6-shim is loaded, it clobbers the native Promise because it doesn't seem to support subclassing, and replaces it with an insidious version with total error suppression:

![image](https://cloud.githubusercontent.com/assets/43438/5430022/f68b9fea-843d-11e4-9f99-21409ce18171.png)

Not sure what the best approach is to solve this.

**Comment by Yaffle** [comment]
@ljharb  I cannot understand this issue, but
```
  var x = Promise.reject(new Error("!"));
  setTimeout(function () {
    x.then(undefined, function (e) {
      console.log(e);
    });
  }, 0);
```
this causes a error message in the log in Chrome. And if replace `setTimeout` with `Promise.resolve(undefined)` - nothing will be printed to the log.
So to simulate same behavior you could use `setTimeout`, right?

**Comment by cscott** [comment]
@Yaffle yes, many utility libraries for Promises (such as http://github.com/cscott/prfun) add a `Promise.prototype.done` method which uses `setTimeout` in exactly this way.

There's also a web spec in-process for triggering the "uncaught promise exception" handler in the same way that there is an "uncaught exception" handler in the web stack for synchronous code.  I don't know if Chrome has exposed this to users yet, though.  If they have, we could hook it.

**Comment by AlastairTaft** [comment]
The try catch around line 1867 is currently eating all my Promise errors (in the `promiseReactionJob` method). All my promise errors are caught and not thrown so it fails silently. I'm going to have to run with a flag to not include the es6-shim when developing in chrome. That also makes me wonder that the shim shouldn't be overriding native chrome promise support.

**Comment by ljharb** [comment]
@AlastairTaft I'm not sure what you mean - that's what native Promises do by spec. All Promises swallow all errors, and the only way to catch them is with a `.catch`.

The additional browser behavior of notifying the console when you have an unhandled rejection is not part of the spec, it's bonus behavior some promises provide.

**Comment by AlastairTaft** [comment]
@ljharb If it's by design then no worries. I'll provide some code though to explain my point a bit better just encase it is a problem.

So I'm doing something like this in my code
    
    var self = this
    // Get the content
    Promise.all([
      api.client.getProfileP(),
      api.client.getInfoSummariesP()
    ])
    .then(function(args){
      var profile = args[0]
        , infoSummaries = args[1]

      self.loadContent_(profile, infoSummaries)
    })
    .catch(function(err){
      throw err
    })

When I hit a run time error in the `self.loadContent_(profile, infoSummaries)` method the `throw err` is swallowed and nothing is output to the console.

**Comment by ljharb** [comment]
Correct. You can never throw an error outside of a Promise - that's how they work everywhere. You can do `setTimeout(function () { throw err; }, 0)` but that's basically your only option.

Handle the error in the right place: inside the promise :-)

**Comment by AlastairTaft** [comment]
Fair enough :)

**Comment by cscott** [comment]
The `prfun` library (and others) have a `Promise #done` method that can
also be helpful in making your thrown errors visible.

**Comment by ertant** [comment]
At least for debug versions I have changed the rejectPromise (line 1913) function like below, otherwise finding an error almost impossible.

    var rejectPromise = function (promise, reason) {
    .....
    if(!reactions.length) {
      console.warn('Uncaught (in promise)', reason);
    }
          triggerPromiseReactions(reactions, reason);
    };
